### ISSUE TITLE
Property interpolation

### ISSUE THREAD

**Issue by philbucher** [KratosMultiphysics/Kratos#825]
I want to simulate a turbine blade that has a varying thickness along its main axis. The idea is the same as if e.g. a temperature dependent youngs modulus for solids or a temperature dependent viscosity for fluids. Each element would have a value (e.g. a specified temperature) and a table that links the temperature with the property that is to be interpolated.
I think this is something that many might use, therefore @KratosMultiphysics/fluid-dynamics @KratosMultiphysics/structural-mechanics @KratosMultiphysics/kratos-core-team 

My idea was to specify some variable (the nondimensional blade length for each element in my case) and  interpolate the Thickness using a `Table`.
This does not work however because when the elements ask the `Properties` for any value, the table values, even though the corresponding table exists in the `Properties`, are not considered. It directly accesses the stored Data (e.g. [here](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/properties.h#L182))
A solution could be to add another `GetValue` function to `Properties` that also passes a pointer/ref to an element/condition. When this function is called, it first checks if there is a `Table` available for the querried variable. If no, then it does the same as what it does atm. However if there is a `Table` available, then it checks the Data of the passed Element/condition for the variable with which the interpolation is done and returns the interpolated value.
Here is some pseudocode to illustrate my idea:
```
GetValue(rVariable, rGeometricalObject)
{
  if (Table exists for rVariable)
  {
    rArgumentVariable = Table.GetPrimaryVariable() // this function would have to be added to Table
    if (rGeometricalObject.Has(rArgumentVariable))
    {
      rArgumentValue = rGeometricalObject.GetValue(rArgumentVariable)
      return Table.GetValue(rArgumentValue)
    }
    else
    {
      KRATOS_ERROR << "Table specified but elemental data is missing\n";
    }
    
  }
  else
  {
    return GetValue(rVariable) // This is what currently happens
  }
}
```
The (very) big disadvantage however is that every element has to access the properties through this function and would have to be changed, so I guess this is not the way to go.

Any opinions or thoughts?

**Comment by jcotela** [comment]
Hi @philbucher. What you are proposing could work, but keep in mind that the element system calculation is quite "deep" in code and all branching and checks we put in there can have an impact in performance.
Regarding your specific proposal, my objection would be that you are somehow assuming that the argument to query the table is given by GeometricalObject::GetValue(). This would not be the case for temperature-based checks, for example, since temperature is a nodal variable (so you would either consult the table at nodes and interpolate the result or pass an interpolated temperature value as argument). We might want to pass `rArgumentValue` in the function call itself.

From the point of view of the fluid, we are currently trying to work out a way to reorganize the internal data of the element that would make it easy to change where elemental data is read from. In our new design, to do what you are proposing would require to write a small auxiliary class that reads variables using your code above and specify (at compile time) that `VISCOSITY` (or whatever) is to be read using your class.

**Comment by loumalouomega** [comment]
I was thinking about this long time ago, but as I don't needed any more I didn't develop my idea. I was thinking something similar to what you propose, but maybe using a flag instead of a check

**Comment by philbucher** [comment]
thx for you comments
@jcotela I see argument with the nodes. But how would you get the `rArgumentValue`? Through a wrapper class? So every element would access the properties through the wrapper class, which knows whether interpolation is necessary or not?
Specifying which variable to use at compile time would limit the usage capabilities quite a bit, wouldn't it?

@loumalouomega could you specify a bit more what you intended to do?

**Comment by jcotela** [comment]
@philbucher my preference would be to pass the `rArgumentValue` as argument in the function call, if required. Just to be clear, my opinion is that the element (not the property, or the GetValue method) should know where to look for its data (even if this includes looking for it in one of several places). My reasoning for this is that there are multiple ways of getting a value (nodal interpolation from historical or non-historical variables, process info, properties, elemental database, tables) and it is hard to generalize in a domain-independent way. For example, going with tables, the argument variable might be different in different applications (lookup by temperature? by position? by distance to some reference surface?). What happens if one ends up with two tables for the same variable. It does not make much sense in your case, but for the sake of discussion consider ending up with temperature-density and distance-density tables. How does the property decide which one to pick?

I think that this information (knowing where to look for each relevant variable) is quite application-dependent in the end. I think your approach is correct, but I'd be inclined to put the logic for it in the domain-specific element itself, rather than in the default Get methods, so that every developer can make an informed decision to look for data on one or more relevant containers.

**Comment by loumalouomega** [comment]
I cannot specify, Ii was thinking about that, but it had an implication in the way that the data in Kratos was stored, therefore I needed to have discussion with @pooyan-dadvand that it never took place.

**Comment by pooyan-dadvand** [comment]
Good question! The first design idea behind the table was exactly as you are asking. And even I arrived to some prototype of the Properties.GetTableValue which takes the node, and give the `VISCOSITY` using the `TEMPERATURE` stored in node. For this reason the table on that time was storing two variables which based. Nevertheless this design shows different problems in practice:
- The interface for using table and value was different (which is not bad by itself) and this difference was exposed in element level which is not good (as @jcotela also mentioned) 
- Having variables stored in table make them restricted in time of using vector components
- There was some constrains in implementation details for tables and their containers.
- The performance penalty for having a sophisticated selection in properties level was constraining any smart and generic implementation.

So finally we have took out the variables from tables leaving them as simple containers and we make some process which applies tables to nodal or elemental data in a sub model part and we call them in each time step when is necessary.
